### PR TITLE
PRVT-6: add call for getLoginMethods

### PR DESCRIPTION
**Pull request by virdesai**
feat : add a call for getLoginMethods and expose clientOptions to users for testing purposes on their end

### CODE REVIEW

File: src/APIBasedAuth.ts [hunk 1]
```diff
@@ -87,7 +85,8 @@ const DEFAULT_STORAGE_KEYS: APIBasedAuth.StorageKeys = {
 
 class APIBasedAuth {
   private client: AxiosInstance;
-  private clientOptions: APIBasedAuth.Config;
+  clientOptions: APIBasedAuth.Config;
```
Comment (virdesai) [inline src/APIBasedAuth.ts:88]: exposing this for testing purposes for clients to ensure they have configured options properly or to access anything they need after the fact

File: src/APIBasedAuth.ts [hunk 2]
```diff
@@ -133,6 +133,18 @@ class APIBasedAuth {
...
+    if (this.loginMethods?.[username]) {
+      return this.loginMethods[username];
+    }
+    const { data } = await this.client.get<
+      APIBasedAuth.LoginMethod[],
+      AxiosResponse<APIBasedAuth.LoginMethod[]>
+    >('/login-methods', { params: { login: username } });
+    this.loginMethods = { [username]: data };
+    return data;
+  }
```
Comment (virdesai) [inline src/APIBasedAuth.ts:146]: not putting any of this into storage as it seems pretty low overhead and we don't want to explode secure storage with tons of { username: loginMethods } values

File: src/APIBasedAuth.ts [hunk 3]
```diff
@@ -10,18 +10,16 @@ const TOKEN_KEYS = [
...
 ];
-const DEFAULT_STORAGE_KEYS: APIBasedAuth.StorageKeys = {
-  accessToken: `${API_AUTH_STORAGE_KEY}.accessToken`,
-  expiresIn: `${API_AUTH_STORAGE_KEY}.expiresIn`,
-  idToken: `${API_AUTH_STORAGE_KEY}.idToken`,
-  refreshToken: `${API_AUTH_STORAGE_KEY}.refreshToken`,
-  session: `${API_AUTH_STORAGE_KEY}.session`,
-  tokenType: `${API_AUTH_STORAGE_KEY}.tokenType`,
-  username: `${API_AUTH_STORAGE_KEY}.username`
-};
```
Comment (smolster) [inline src/APIBasedAuth.ts:21]: :( I thought this was much more readable.

Also, looks like we lost the `APIBasedAuth.StorageKeys` type?

File: src/APIBasedAuth.ts [hunk 4]
```diff
@@ -133,6 +133,18 @@ class APIBasedAuth {
...
+  public async getLoginMethods(username: string) {
+    if (this.loginMethods?.[username]) {
+      return this.loginMethods[username];
+    }
+    const { data } = await this.client.get<
+      APIBasedAuth.LoginMethod[],
+      AxiosResponse<APIBasedAuth.LoginMethod[]>
+    >('/login-methods', { params: { login: username } });
+    this.loginMethods = { [username]: data };
+    return data;
```
Comment (smolster) [inline src/APIBasedAuth.ts:145]: As-is, I don't think this code allows a client to trigger a fresh request for a username without completely re-instantiating the class -- right?

It seems like most clients would come with their own opinions about caching network results. `ClientUI` does, for example.

Maybe we should skip the local cache here, and just let the client cache results? I feel like most pure API SDKs I've interacted don't seem to do internal caching.

File: src/APIBasedAuth.ts (hunk 1 above)
Comment (smolster) [reply src/APIBasedAuth.ts:88]: Should we mark this as `readonly`?

File: src/APIBasedAuth.ts (hunk 3 above)
Comment (virdesai) [reply src/APIBasedAuth.ts:21]: the type is still there but I'll revert to the more readable one. this format just allowed us to expand to allow more storage if desired in the future

**Review by smolster — APPROVED** [review]
+1 to alejandro's comment. But, fine with me to do that in a subsequent PR.

**Review by jkdowdle — APPROVED** [review]

**Comment by loscm** [comment]
Requested review from a random `@lifeomic/security` team member, @cfbevan, due to term: **expose**.

This review is not blocking and is for broader awareness.  Consider if this change requires deeper security review and ask when it is necessary.

- [Automation](https://github.com/lifeomic/pr-security-flag-scanner/blob/master/src/lambdas/pr-security-flag-scanner/index.ts) | [#pr-flagged-sec-review](https://lifeomic.slack.com/messages/pr-flagged-sec-review)

**Pull request merged**